doc/developer/design/20230119_ir_normalization.md

```diff
@@ -0,0 +1,147 @@
+# Intermediate Representation Normalization
+
+## Summary
+
+<!--
+// Brief, high-level overview. A few sentences long.
+// Be sure to capture the customer impact - framing this as a release note may be useful.
+-->
+
+Introduce normalizing transformations to some of HIR, MIR, and LIR that bring ASTs to a normal form with certain predictable properties.
+The intent is to introduce *predictability* to the representation of an AST, so that transformations, explanations, and other moments where Materialize and users interact with an expression AST do not have to worry or reason about as many equivalent alternatives.
+
+## Goals
```

Comment:
We could add some discussion of even higher level goals, i.e., desired effects that would be consequences of the goals listed so far. For example:
1. We want to make CSE more effective. This seems like a very clear and universal benefit of normalization.
2. "optimizing transforms that assume normalized input have an easier job of recognizing the input patterns that they want to transform" (taking this from @aalexandrov's comment below). I think this needs some discussion, because it only works if the normalization is reliable. For example:
    - If `Union` inputs are sorted simply by the `Ord` of `MirRelationExpr`, and (let's say) the complicated pattern matching in unique key detection of `Union` would rely on this order, then we would risk breaking that pattern matching if the canonical order changes. Such an `Ord` change could happen from, for example, https://github.com/MaterializeInc/materialize/pull/16240.
    - `extract_equal_or_both_null` used to match only one input order, because its input was canonicalized. However,  it still broke suddenly. This happened when the input order flipped during the "make AND/OR variadic" work changing `MirScalarExpr`'s `Ord` by varying some `MirScalarExpr` enum variants. Then I made it match in both directions to avoid any future regressions if the canonical order changes again somehow.
    - [This pattern match](https://github.com/MaterializeInc/materialize/blob/b15c99c3d915bdc33620d06f9dccbf1a82d5eeab/src/expr/src/scalar/mod.rs#L1003) relying on [this canonicalization](https://github.com/MaterializeInc/materialize/blob/b15c99c3d915bdc33620d06f9dccbf1a82d5eeab/src/expr/src/scalar/mod.rs#L977-L983), which again depends on the `Ord` of `MirScalarExpr`.

Maybe we can still have 2., but with the caveat that those canonicalizations that depend on `Ord` are _not_ to be relied upon by patterns. (This still leaves other canonicalizations ok to be relied upon. E.g., Map fusion would ensure that when `ReductionPushdown` looks for a Map on a `Reduce` input, there is no need to look for multiple Maps.)